### PR TITLE
Fix issues found while experimenting with `-flto`

### PR DESCRIPTION
**Pull request by evan-charmworks**

### CODE REVIEW

**Comment by evan-charmworks** [comment]
With this PR, adding `-flto` to the end of a `./build all-test` command completes successfully with no warnings on GCC, and with Clang while Fortran is disabled (since clang and gfortran don't understand each other's intermediate representation).

All tests pass except for `tlsglobals-f90` which crashes, but I have reproduced the bug with a modified test case on conventional MPI without any Charm/AMPI code or CmiTLS functionality and will document it.

Building `all-test` with LTO takes a very long time since it optimizes the entire runtime for each binary created, and for a case like integration testing it would be better to test LTO with `--build-shared -charm-shared`, which also coincidentally avoids the `tlsglobals-f90` crash. Though, splitting the code into shared objects reduces the extent of how LTO can be applied, and that is compounded by how many small shared objects we generate.

**Comment by evan-charmworks** [comment]
I think I've scorched the earth of compiler warnings enough for one PR. Ready for review.

File: src/ck-ldb/ckgraph.C [hunk 1]
```diff
@@ -62,8 +62,8 @@ ObjGraph::ObjGraph(BaseLB::LDStats *stats) {
       from = stats->getHash(commData.sender);
       to = stats->getHash(commData.receiver.get_destObj());
 
-      vertices[from].sendToList.push_back(Edge(to, commData.messages, commData.bytes));
-      vertices[to].recvFromList.push_back(Edge(from, commData.messages, commData.bytes));
+      vertices[from].sendToList.push_back(CkEdge(to, commData.messages, commData.bytes));
```
Comment (stwhite91) [inline src/ck-ldb/ckgraph.C:65]: emplace_back?

File: src/ck-core/ckrdma.C [hunk 2]
```diff
@@ -2064,11 +2064,15 @@ void CkPostBufferInternal(void *destBuffer, size_t destSize, int tag) {
 void CkPostNodeBufferInternal(void *destBuffer, size_t destSize, int tag) {
 
   // check if tag exists in posted req node table
+  CmiLock(CksvAccess(_nodeZCPostReqLock));
```
Comment (stwhite91) [inline src/ck-core/ckrdma.C:2067]: Just curious, what compiler / options pointed this out with LTO?

File: src/ck-ldb/ckgraph.C (hunk 1 above)
Comment (evan-charmworks) [reply src/ck-ldb/ckgraph.C]: Nice, done.

File: src/ck-core/ckrdma.C (hunk 2 above)
Comment (evan-charmworks) [reply src/ck-core/ckrdma.C:2067]: I was building with GCC on x86_64 Linux, but the issue manifested as a crash in one of the ZC tests or examples and not a diagnostic message. The crash happened in the implementation of `std::unordered_map` and traced back to these `find()` calls.

File: examples/charm++/load_balancing/stencil3d/stencil3d.C [hunk 3]
```diff
@@ -324,7 +324,7 @@ class Stencil: public CBase_Stencil {
         work = 10.0;
 
 #ifndef _MSC_VER
```
Comment (stwhite91) [inline examples/charm++/load_balancing/stencil3d/stencil3d.C:326]: I would be in favor of just removing these, rather than commenting them out

Comment (evan-charmworks) [reply examples/charm++/load_balancing/stencil3d/stencil3d.C]: Done

File: src/ck-core/ckmulticast.C [hunk 4]
```diff
@@ -46,7 +46,7 @@ typedef unsigned char byte;
  * An instance of this class is stored in every mCastEntry object making it possible
  * to track redn operations on a per section basis all along the spanning tree.
  */
-class reductionInfo {
+class multicastRedInfo {
```
[the remark this answers is not in the document]
Comment (evan-charmworks) [reply src/ck-core/ckmulticast.C]: `sectionRedInfo` makes sense.

In the Charm repository, `redInfo` and `redMsg` outnumber `rednInfo` and `rednMsg`, while `rednMgr` is more frequent than `redMgr`.

**Review by stwhite91 — APPROVED** [review]

**Review by rbuch — APPROVED** [review]

**Pull request merged**